packages/go/go-helpers.ts

```diff
@@ -112,72 +134,222 @@ class GoWrapper {
+ * @param modulePath The path possibly containing a `go.mod` file
+ * @param opts `execa` options (`cwd`, `env`, `stdio`, etc)
+ * @param workPath The path to the project to be built
+ * @returns An initialized `GoWrapper` instance
+ */
+export async function createGo({
+  modulePath,
+  opts = {},
+  workPath,
+}: CreateGoOptions): Promise<GoWrapper> {
+  // parse the `go.mod`, if exists
+  let goPreferredVersion;
+  if (modulePath) {
+    goPreferredVersion = await parseGoModVersion(modulePath);
+  }
+
+  // default to newest (first) supported go version
+  const goSelectedVersion =
+    goPreferredVersion || Array.from(versionMap.values())[0];
+
+  const env = cloneEnv(process.env, opts.env);
+  const { PATH } = env;
+  const { platform } = process;
+  const goGlobalCacheDir = join(
+    goGlobalCachePath,
+    `${goSelectedVersion}_${platform}_${process.arch}`
+  );
+  const goCacheDir = join(workPath, cacheDir);
+
+  if (goPreferredVersion) {
+    debug(`Preferred go version ${goPreferredVersion} (from go.mod)`);
     env.GO111MODULE = 'on';
+  } else {
+    debug(
+      `Preferred go version ${goSelectedVersion} (latest from version map)`
+    );
+  }
+
+  const setGoEnv = async (goDir: string | null) => {
+    if (platform !== 'win32' && goDir === goGlobalCacheDir) {
+      debug(`Symlinking ${goDir} -> ${goCacheDir}`);
+      await mkdirp(dirname(goCacheDir));
+      await remove(goCacheDir);
+      await symlink(goDir, goCacheDir);
+      goDir = goCacheDir;
+    }
+    env.GOROOT = goDir || undefined;
+    env.PATH = goDir ? `${join(goDir, 'bin')}${delimiter}${PATH}` : PATH;
+  };
+
+  // try each of these Go directories looking for the version we need
+  const goDirs = {
+    'local cache': goCacheDir,
+    'global cache': goGlobalCacheDir,
+    'system PATH': null,
+  };
+
+  for (const [label, goDir] of Object.entries(goDirs)) {
+    try {
+      const goBinDir = goDir && join(goDir, 'bin');
+      if (goBinDir && !(await pathExists(goBinDir))) {
+        debug(`Go not found in ${label}`);
+        continue;
+      }
+
+      env.GOROOT = goDir || undefined;
+      env.PATH = goBinDir || PATH;
+
+      const { stdout } = await execa('go', ['version'], { env });
+      const { minor, short, version } = parseGoVersionString(stdout);
+
+      if (minor < GO_MIN_VERSION) {
+        debug(`Found go ${version} in ${label}, but version is unsupported`);
+      }
+      if (version === goSelectedVersion || short === goSelectedVersion) {
+        debug(`Selected go ${version} (from ${label})`);
+
+        await setGoEnv(goDir);
+        return new GoWrapper(env, opts);
+      } else {
+        debug(`Found go ${version} in ${label}, but need ${goSelectedVersion}`);
+      }
+    } catch {
+      debug(`Go not found in ${label}`);
+    }
   }
-  await createGoPathTree(goPath, platform, arch);
+
+  // we need to download and cache the desired `go` version
+  await download({
+    dest: goGlobalCacheDir,
+    version: goSelectedVersion,
+  });
```

Comment:
**suggestion (non-blocking):** log message when downloading

When I hit a Go api endpoint and Go needs to be downloaded, I can wait many seconds before seeing any feedback. Can we log a message here saying that Go is being downloaded?